File: SLL/SLL.c
#include <stdio.h>
#include <stdlib.h>


struct SLL
{
    int data;
    struct SLL *next;
};

struct SLL *first, *last = NULL;

void insertAtBeg(int element)
{
    struct SLL *NewNode;
    NewNode = (struct SLL*)malloc(sizeof(struct SLL));
    NewNode->next = NULL;
    NewNode->data = element;

    if(first == NULL)
        first = last = NewNode;

    else
    {
        NewNode->next = first;
        first = NewNode;
    }
    printf("%d WAS INSERTED AT THE BEGINNING\n",first->data);
}

void insertAtEnd(int element)
{
    struct SLL *NewNode;
    NewNode = (struct SLL*)malloc(sizeof(struct SLL));
    NewNode->next = NULL;
    NewNode->data = element;

    if(first == NULL)
        first = last = NewNode;

    else
    {
        last->next = NewNode;
        last = NewNode;
    }

    printf("%d WAS INSERTED AT THE END\n",last->data);
}

void insertAtPosition(int element, int pos)
{
    struct SLL *NewNode, *temp;
    int i;
    NewNode = (struct SLL*)malloc(sizeof(struct SLL));
    NewNode->next = NULL;
    NewNode->data = element;

    if(first == NULL || pos <= 0)
        printf("INVALID INSERTION!\n");

    else if(pos == 1)
        insertAtBeg(element);

    else
    {
        temp = first;
        for(i = 1; i < pos-1 && temp != NULL; i++)
            temp = temp->next;

        if(temp == NULL)
            insertAtEnd(element);

        else
        {
            NewNode->next = temp->next;
            temp->next = NewNode;
            printf("%d WAS INSERTED AT POSITION %d\n",NewNode->data, pos);
        }
    }
}

void deletionFromBeg()
{
    struct SLL *temp;
    if(first == NULL)
        printf("EMPTY LIST!\n");

    else if(first->next == NULL)
    {
        temp = first;
        printf("%d IS GOING TO BE DELETED\n",temp->data);
        first = last = NULL;
        free(temp);
    }

    else
    {
        temp = first;
        first = first->next;
        printf("%d IS GOING TO BE DELETED\n",temp->data);
        free(temp);
    }
}

void deletionFromEnd()
{
    struct SLL *temp;
    if(first == NULL)
        printf("EMPTY LIST!\n");

    else if(first->next == NULL)
    {
        temp = first;
        printf("%d IS GOING TO BE DELETED\n",temp->data);
        first = last = NULL;
        free(temp);
    }

    else
    {
        temp = first;
        while(temp->next != last)
        {
            temp = temp->next;
        }
        last = temp;
        temp = temp->next;
        printf("%d IS GOING TO BE DELETED\n",temp->data);
        free(temp);
        last->next = NULL;
    }
}

void deletionFromPosition(int pos)
{
    struct SLL *temp, *tempp;
    int i;
    if(first == NULL || pos<=0)
        printf("INVALID DELETION!\n");

    else if(first->next ==  NULL || pos == 1)
    {
        deletionFromBeg();
    }

    else
    {
        temp = first;

        for(i = 1; i < pos-1; i++)
            temp = temp->next;

        if(temp->next == last)
            deletionFromEnd();

        else
        {
            tempp = temp->next;
            temp->next = tempp->next;
            printf("%d IS GOING TO BE DELETED\n",tempp->data);
            free(tempp);
        }
    }

}

void display()
{
    struct SLL *temp;
    if(first == NULL)
        printf("EMPTY LIST!\n");
    else
    {
        temp = first;
        while(temp != NULL)
        {
            printf("%d -> ",temp->data);
            temp = temp->next;
        }
        printf("NULL\n");
    }
}


void search(int key)
{
    struct SLL *temp;
    int flag = 0;

    if(first == NULL)
        printf("EMPTY LIST!\n");

    else
    {
        temp = first;
        while(temp != NULL)
        {
            if(temp->data == key)
            {
                printf("Element Found!\n");
                flag = 1;
            }
            temp = temp->next;

        }
        if(flag == 0)
            printf("SEARCH UNSUCCESSFUL!\n");
    }
}
int main()
{
    int choice, subchoice, element, pos;

    while (1) 
    {
        printf("\n--- MENU ---\n");
        printf("1. Insert\n2. Delete\n3. Search\n4. Display\n5. Exit\n");

        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) 
        {
        case 1:
            printf("Insert:\n1. At Beginning\n2. At End\n3. At Specific Position\n");

            printf("Enter your choice: ");
            scanf("%d", &subchoice);

            printf("Enter element to insert: ");
            scanf("%d", &element);

            if (subchoice == 1)
                insertAtBeg(element);

            else if (subchoice == 2)
                insertAtEnd(element);

            else if (subchoice == 3) 
            {
                printf("Enter position: ");
                scanf("%d", &pos);
                insertAtPosition(element, pos);
            }
            else
                printf("Invalid insert choice!\n");
            break;

        case 2:
            printf("Delete:\n1. From Beginning\n2. From End\n3. From Specific Position\n");

            printf("Enter your choice: ");
            scanf("%d", &subchoice);

            if (subchoice == 1)
                deletionFromBeg();

            else if (subchoice == 2)
                deletionFromEnd();

            else if (subchoice == 3) 
            {
                printf("Enter position: ");
                scanf("%d", &pos);
                deletionFromPosition(pos);
            } 
            else
                printf("Invalid delete choice!\n");
            break;

        case 3:
            printf("Enter element to search: ");
            scanf("%d", &element);
            search(element);
            break;

        case 4:
            display();
            break;

        case 5:
            exit(0);

        default:
            printf("Invalid choice. Try again.\n");
        }
    }

    return 0;
}

File: LinkedLists/SLL/SLL.py
def create_node(data):
    return {'data': data, 'next': None}

first = None
last = None

def InsertAtBeg(element):

    global first, last
    new_node = create_node(element)

    if first is None:
        first = last = new_node

    else:
        new_node['next'] = first
        first = new_node

    print(f"{element} WAS INSERTED AT THE BEGINNING")

def InsertAtEnd(element):
    global first, last
    new_node = create_node(element)

    if first is None:
        first = last = new_node

    else:
        last['next'] = new_node
        last = new_node

    print(f"{element} WAS INSERTED AT THE END")

def InsertAtPosition(element, pos):
    global first

    if first is None or pos <= 0:
        print("INVALID INSERTION!")

    elif pos == 1:
        InsertAtBeg(element)

    else:
        new_node = create_node(element)
        temp = first
        i = 1

        while i < pos - 1 and temp is not None:
            temp = temp['next']
            i += 1

        if temp is None:
            InsertAtEnd(element)

        else:
            new_node['next'] = temp['next']
            temp['next'] = new_node
            print(f"{element} WAS INSERTED AT POSITION {pos}")

def DeleteFromBeg():
    global first, last

    if first is None:
        print("EMPTY LIST!")

    elif first['next'] is None:
        print(f"{first['data']} IS GOING TO BE DELETED")
        first = last = None

    else:
        print(f"{first['data']} IS GOING TO BE DELETED")
        first = first['next']

def DeleteFromEnd():
    global first, last
    if first is None:
        print("EMPTY LIST!")

    elif first['next'] is None:
        print(f"{first['data']} IS GOING TO BE DELETED")
        first = last = None

    else:
        temp = first
        while temp['next'] != last:
            temp = temp['next']

        print(f"{last['data']} IS GOING TO BE DELETED")
        last = temp
        last['next'] = None

def DeleteFromPosition(pos):
    global first, last
    if first is None or pos <= 0:
        print("INVALID DELETION!")

    elif first['next'] is None or pos == 1:
        DeleteFromBeg()

    else:
        temp = first

        for i in range(1, pos - 1):
            if temp is None or temp['next'] is None:
                print("INVALID POSITION!")
                return
            
            temp = temp['next']

        if temp['next'] == last:
            DeleteFromEnd()

        elif temp['next'] is not None:
            print(f"{temp['next']['data']} IS GOING TO BE DELETED")
            temp['next'] = temp['next']['next']

        else:
            print("INVALID POSITION!")

def display():
    global first

    if first is None:
        print("EMPTY LIST!")
        
    else:
        temp = first
        while temp is not None:
            print(f"{temp['data']} -> ", end="")
            temp = temp['next']
        print("NULL")

def search(key):
    global first

    if first is None:
        print("EMPTY LIST!")

    else:
        temp = first
        found = False

        while temp is not None:
            if temp['data'] == key:
                found = True
                break

            temp = temp['next']

        print("Element Found!" if found else "SEARCH UNSUCCESSFUL!")

def main():
    while True:
        print("\n--- MENU ---")
        print("1. Insert\n2. Delete\n3. Search\n4. Display\n5. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            print("Insert:\n1. At Beginning\n2. At End\n3. At Specific Position")
            subchoice = int(input("Enter your choice: "))
            element = int(input("Enter element to insert: "))

            if subchoice == 1:
                InsertAtBeg(element)

            elif subchoice == 2:
                InsertAtEnd(element)

            elif subchoice == 3:
                pos = int(input("Enter position: "))
                InsertAtPosition(element, pos)

            else:
                print("Invalid insert choice!")

        elif choice == 2:
            print("Delete:\n1. From Beginning\n2. From End\n3. From Specific Position")
            subchoice = int(input("Enter your choice: "))

            if subchoice == 1:
                DeleteFromBeg()

            elif subchoice == 2:
                DeleteFromEnd()

            elif subchoice == 3:
                pos = int(input("Enter position: "))
                DeleteFromPosition(pos)
                
            else:
                print("Invalid delete choice!")

        elif choice == 3:
            key = int(input("Enter element to search: "))
            search(key)

        elif choice == 4:
            display()

        elif choice == 5:
            print("Exiting program.")
            break

        else:
            print("Invalid choice. Try again.")


if __name__ == "__main__":
    main()
